### PR TITLE
Gen 8 Evolution Framework

### PR DESCRIPTION
**Pull request by AgustinGDLV**
Added a framework to create "special evolutions" that occur without leveling up or using an item and implemented Galarian Farfetch'd's evolution, as well as functional placeholders for Kubfu and Galarian Yamask.

## Description
Added a TrySpecialEvolution function to battle_main.c. Currently, this is called in place of TryEvolvePokemon before returning to overworld and attempts to perform a special evolution on every party mon.  A static u8 "gTriedEvolving" is used in case an evolution occurs but is cancelled so as to prevent the function from trying to evolve that mon again. The original TryEvolvePokemon is called after, but will go right to returning to the overworld if no mons have levelled up.

Added a script command "tryspecialevo" that takes an evolution method to attempt to perform as an argument. The command attempts to perform the specified evolution method on every party mon. Implemented like other vanilla script commands, wasn't sure if I should've done it like "settotemboost" or not.

Added functionality for Galarian Farfetch'd's evolution method using a bitfield (u16 gPartyCriticalHits) that counts how many times each player party mon has crit in a single battle. Janky bit arithmetic.

Added placeholder evolution methods for Kubfu and Galarian Yamask. Currently, Kubfu will evolve into Single Strike or Rapid Strike form if you an event calls "tryspecialevo EVO_DARK_SCROLL" or "tryspecialevo EVO_WATER_SCROLL," respectively. G-Yamask will evolve if an event (you can place a trigger under a bridge) calls "tryspecialevo EVO_RUNERIGUS" and the Yamask's HP is 49 below max. These can be modified fairly easily, all the functionality is in GetEvolutionTargetSpecies.

Note: For GetEvolutionTargetSpecies, party slot is passed instead for evolutionItem for special battle evolutions and similarly with evoMethod for special overworld evolutions.

I haven't performed very thorough testing yet, so there are probably some bugs to squash.

## **Discord Contact Info**
Agustin#1522

### CODE REVIEW

File: src/wild_encounter.c [hunk 1]
```diff
@@ -271,7 +271,7 @@ static u8 ChooseWildMonLevel(const struct WildPokemon *wildPokemon)
     u8 max;
     u8 range;
     u8 rand;
-
+
```
Comment (LOuroboros) [inline src/wild_encounter.c:274]:
```suggestion

```

File: src/scrcmd.c [hunk 2]
```diff
@@ -2331,3 +2332,20 @@ bool8 ScrCmd_warpwhitefade(struct ScriptContext *ctx)
...
+    {
+        u16 targetSpecies = GetEvolutionTargetSpecies(&gPlayerParty[i], EVO_MODE_OVERWORLD_SPECIAL, evoMethod, SPECIES_NONE);
+        if (targetSpecies != SPECIES_NONE)
+        {
+            gCB2_AfterEvolution = CB2_ReturnToField;
+            BeginEvolutionScene(&gPlayerParty[i], targetSpecies, TRUE, i);
+        }   
+    }
+    return TRUE;
+}
```
Comment (LOuroboros) [inline src/scrcmd.c:2351]: Add a new, empty line here.

File: src/pokemon.c [hunk 3]
```diff
@@ -6554,6 +6554,41 @@ u16 GetEvolutionTargetSpecies(struct Pokemon *mon, u8 mode, u16 evolutionItem, u
             }
         }
         break;
+    //Battle evolution without leveling; party slot is being passed into the evolutionItem arg.
```
Comment (LOuroboros) [inline src/pokemon.c:6557]:
```suggestion
    // Battle evolution without leveling; party slot is being passed into the evolutionItem arg.
```

File: src/pokemon.c [hunk 4]
```diff
@@ -6554,6 +6554,41 @@ u16 GetEvolutionTargetSpecies(struct Pokemon *mon, u8 mode, u16 evolutionItem, u
...
+            switch (gEvolutionTable[species][i].method)
+            {
+            case EVO_THREECRITS:
+                if (((gPartyCriticalHits >> evolutionItem*2) & 3) == 3)
+                    targetSpecies = gEvolutionTable[species][i].targetSpecies;
+                break;
+            }
+        }
+        break;
+    //Overworld evolution without leveling; evolution method is being passed into the evolutionItem arg.
```
Comment (LOuroboros) [inline src/pokemon.c:6570]:
```suggestion
    // Overworld evolution without leveling; evolution method is being passed into the evolutionItem arg.
```

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -1922,6 +1924,18 @@ static void Cmd_critcalc(void)
     else
         gIsCriticalHit = FALSE;
 
+    //Counter for Sirfetchd evolution method (caps at 3 per party mon).
```
Comment (LOuroboros) [inline src/battle_script_commands.c:1927]:
```suggestion
    // Counter for Sirfetchd evolution method (caps at 3 per party mon).
```

File: data/script_cmd_table.inc [hunk 6]
```diff
@@ -227,6 +227,7 @@ gScriptCmdTable::
 	.4byte ScrCmd_warpwhitefade             @ 0xe0
 	.4byte ScrCmd_buffercontestname         @ 0xe1
 	.4byte ScrCmd_bufferitemnameplural      @ 0xe2
+	.4byte ScrCmd_tryspecialevo		        @ 0xe3
```
Comment (LOuroboros) [inline data/script_cmd_table.inc:230]:
```suggestion
	.4byte ScrCmd_bufferitemnameplural      @ 0xe2
	.4byte ScrCmd_tryspecialevo                   @ 0xe3
```

File: include/battle.h [hunk 7]
```diff
@@ -926,5 +926,6 @@ extern bool8 gHasFetchedBall;
 extern u8 gLastUsedBall;
 extern u16 gLastThrownBall;
 extern bool8 gSwapDamageCategory; // Photon Geyser, Shell Side Arm, Light That Burns the Sky
+extern u16 gPartyCriticalHits; // Bitfield to track the critical hits of each party member, up to 3. Used for Sirfetchd's evolution method.
```
Comment (LOuroboros) [inline include/battle.h:929]:
```suggestion
extern u16 gPartyCriticalHits; // Bitfield to track the critical hits of each party member, up to 3. Used for EVO_THREE_CRITS.
```

File: src/battle_main.c [hunk 8]
```diff
@@ -5190,10 +5203,30 @@ static void TryEvolvePokemon(void)
     gBattleMainFunc = ReturnFromBattleToOverworld;
 }
 
+static void TrySpecialEvolution(void) //Not related to the tryspecialevo script command.
```
Comment (LOuroboros) [inline src/battle_main.c:5206]:
```suggestion
static void TrySpecialEvolution(void) // Not related to the tryspecialevo script command.
```

File: include/constants/pokemon.h [hunk 9]
```diff
@@ -383,14 +383,20 @@
 #define EVO_SPECIFIC_MAP                  32     // Pokémon levels up on specified map
 #define EVO_LEVEL_NATURE_AMPED            33     // Pokémon reaches the specified level, it has a Hardy, Brave, Adamant, Naughty, Docile, Impish, Lax, Hasty, Jolly, Naive, Rash, Sassy, or Quirky nature.
 #define EVO_LEVEL_NATURE_LOW_KEY          34     // Pokémon reaches the specified level, it has a Lonely, Bold, Relaxed, Timid, Serious, Modest, Mild, Quiet, Bashful, Calm, Gentle, or Careful nature.
+#define EVO_THREECRITS                    35     // Pokémon performs three critical hits in one battle
```
Comment (LOuroboros) [inline include/constants/pokemon.h:386]:
```suggestion
#define EVO_THREE_CRITS                   35     // Pokémon performs three critical hits in one battle
```

File: src/data/pokemon/evolution.h [hunk 10]
```diff
@@ -507,13 +507,13 @@ const struct Evolution gEvolutionTable[NUM_SPECIES][EVOS_PER_MON] =
     [SPECIES_PONYTA_GALARIAN]       = {{EVO_LEVEL, 40, SPECIES_RAPIDASH_GALARIAN}},
     [SPECIES_SLOWPOKE_GALARIAN]     = {{EVO_ITEM, ITEM_NONE, SPECIES_SLOWBRO_GALARIAN},
                                        {EVO_ITEM, ITEM_NONE, SPECIES_SLOWKING_GALARIAN}},
-    [SPECIES_FARFETCHD_GALARIAN]    = {{EVO_LEVEL, 0, SPECIES_SIRFETCHD}},
+    [SPECIES_FARFETCHD_GALARIAN]    = {{EVO_THREECRITS, 0, SPECIES_SIRFETCHD}},
```
Comment (LOuroboros) [inline src/data/pokemon/evolution.h:510]:
```suggestion
    [SPECIES_FARFETCHD_GALARIAN]    = {{EVO_THREE_CRITS, 0, SPECIES_SIRFETCHD}},
```

File: src/pokemon.c [hunk 11]
```diff
@@ -6554,6 +6554,41 @@ u16 GetEvolutionTargetSpecies(struct Pokemon *mon, u8 mode, u16 evolutionItem, u
             }
         }
         break;
+    //Battle evolution without leveling; party slot is being passed into the evolutionItem arg.
+    case EVO_MODE_BATTLE_SPECIAL:
+        for (i = 0; i < EVOS_PER_MON; i++)
+        {
+            switch (gEvolutionTable[species][i].method)
+            {
+            case EVO_THREECRITS:
```
Comment (LOuroboros) [inline src/pokemon.c:6563]:
```suggestion
            case EVO_THREE_CRITS:
```

**Review by LOuroboros — CHANGES_REQUESTED** [review]
There's a lot of syntax corrections to make.

File: src/pokemon.c [hunk 12]
```diff
@@ -6554,6 +6554,41 @@ u16 GetEvolutionTargetSpecies(struct Pokemon *mon, u8 mode, u16 evolutionItem, u
...
+                if (((gPartyCriticalHits >> evolutionItem*2) & 3) == 3)
```
Comment (LOuroboros) [inline src/pokemon.c:6564]:
```suggestion
                if (((gPartyCriticalHits >> evolutionItem * 2) & 3) == 3)
```

**Comment by LOuroboros** [comment]
Oh yeah, you submitted this PR to the wrong branches btw.

Master is not meant to receive any PRs.

The battle specific parts of this PR should be sent to the battle_engine, and the overworld scripting ones should be sent to the pokemon_expansion, imo.

File: src/battle_script_commands.c [hunk 13]
```diff
@@ -1922,6 +1924,18 @@ static void Cmd_critcalc(void)
     else
         gIsCriticalHit = FALSE;
 
+    // Counter for Sirfetchd evolution method (caps at 3 per party mon).
+    partySlot = gBattlerPartyIndexes[gBattlerAttacker];
+    if (gIsCriticalHit && GetBattlerSide(gBattlerAttacker) == 0
```
Comment (LOuroboros) [inline src/battle_script_commands.c:1929]:
```suggestion
    if (gIsCriticalHit && GetBattlerSide(gBattlerAttacker) == B_SIDE_PLAYER
```

**Comment by AsparagusEduardo** [comment]
#2113 #2112

**Pull request closed without merging**